src/understand-tidb/parser.md

```diff
@@ -1,7 +1,76 @@
 # Parser
 
-The code for the parser lives in the [pingcap/parser](https://github.com/pingcap/parser/) repo.
+Parser is responsible for interpreting a SQL string into an abstract syntax tree(AST), which is more structural and easier to process. AST can be used for preprocess, syntactic analysis, and so on.
 
-The main file is `parser.y` which is used by [goyacc](https://pkg.go.dev/golang.org/x/tools/cmd/goyacc) to generate `parser.go`.
+The code lives in the [pingcap/parser](https://github.com/pingcap/parser) repo.
 
-After making changes in the parser use `make` to update the parser and follow [this procedure](https://github.com/pingcap/parser/blob/master/docs/update-parser-for-tidb.md) to have TiDB use your copy of the parser. This can be a local copy or a copy on GitHub. This allows you to make changes to both the parser and TiDB and test them together.
\ No newline at end of file
+## Understand Parser
+
+Parser is generated by a parser generator named [yacc](https://github.com/cznic/parser). It takes the grammar file `parser.y` as the input and output the source code file `parser.go`, which is the real parser imported by TiDB. Thus, the core file is `parser.y` because when the SQL syntax changes, most of the changes take place in `parser.y`.
+
+In case you are unfamiliar with yacc, some concepts are listed here:
+
+- **Terminal Symbol** is also known as "token". When a SQL string reaches parser, the first step is to tokenize them into an array of tokens. For example, `"SELECT * FROM t"` is tokenized to `[selectKeyword, '*', fromKeyword, identifier(t)]` by [lexer.Lex()](https://github.com/pingcap/parser/blob/ae8dec26d777272d56288fa4c9d4264ce0d31a03/lexer.go#L178).
+- **Non-terminal Symbol** is a syntatic variable, which can represent a group of terminal/non-terminal symbols.
+- **Grammar Rule** sepcifies which symbols can replace which other non-terminal symbol.
+- **Semantic Action** defines how an AST node is constructed.
+
+An example of a grammar rule is as follows:
+
+```
+AlterDatabaseStmt:
+	"ALTER" DatabaseSym DBName DatabaseOptionList
+	{
+		$$ = &ast.AlterDatabaseStmt{
+			Name:                 $3,
+			AlterDefaultDatabase: false,
+			Options:              $4.([]*ast.DatabaseOption),
+		}
+	}
+```
+
+- `AlterDatabaseStmt` is a non-terminal symbol because there is no such token.
+- `"ALTER"` is a terminal symbol.
+- `DatabaseSym`, `DBName`, `DatabaseOptionList` are non-terminal symbols that are defined in other rules.
+- The psudo-code in brackets is the semantic action. It means an AST node `ast.AlterDatabaseStmt` will be constructed when the rule is reduced by the parser. Note that a dollar character `$` followed by a number represents the binded Golang value previously(in other rules), where the number is the index of symbol in rule. `$$` represents current binding value. After goyacc substitution, this code snippet will be valid Golang code.
```

Comment:
```suggestion
* `AlterDatabaseStmt` is a non-terminal symbol because there is no such token.
* `"ALTER"` is a terminal symbol.
* `DatabaseSym`, `DBName`, `DatabaseOptionList` are non-terminal symbols that are defined in other rules.
* The pseudo-code in brackets is the semantic action. It means an AST node `ast.AlterDatabaseStmt` will be constructed when the rule is reduced by the parser. Note that a dollar character `$` followed by a number represents the binding Golang value previously (in other rules), where the number is the index of symbol in rule (1-based). `$$` represents current binding value. After goyacc substitution, this code snippet will be valid Golang code.
```